algorithms/kernel/kernel_function/kernel_function_rbf_dense_default_impl.i

```diff
@@ -174,72 +156,142 @@ services::Status KernelImplRBF<defaultDense, algorithmFPType, cpu>::computeInter
+                PRAGMA_VECTOR_ALWAYS
+                for (size_t j = 0; j < nFeatures; j++)
+                {
+                    sqrDataA2[i] += dataA2[(i - startRow) * nFeatures + j] * dataA2[(i - startRow) * nFeatures + j];
+                }
             }
-        }
-        for (size_t i = 0; i < nVectors1; i++)
-        {
-            for (size_t k = 0; k < nVectors2; k++)
+        });
+
+        daal::threader_for(blockNum1, blockNum1, [&](const size_t iBlock) {
+            const size_t numRowsInBlock = (iBlock != blockNum1 - 1) ? blockSize : nVectors1 - iBlock * blockSize;
+            const size_t startRow       = iBlock * blockSize;
+            const size_t finishRow      = startRow + numRowsInBlock;
+
+            algorithmFPType * dataRLocal = dataR + startRow * nVectors2;
+
+            for (size_t i = startRow; i < finishRow; ++i)
             {
-                dataR[i * nVectors2 + k] += (sqrDataA1[i] + sqrDataA2[k]);
-                dataR[i * nVectors2 + k] *= coeff;
-                if (dataR[i * nVectors2 + k] < Math<algorithmFPType, cpu>::vExpThreshold())
+                PRAGMA_IVDEP
+                PRAGMA_VECTOR_ALWAYS
+                for (size_t k = 0; k < nVectors2; ++k)
                 {
-                    dataR[i * nVectors2 + k] = Math<algorithmFPType, cpu>::vExpThreshold();
+                    dataRLocal[(i - startRow) * nVectors2 + k] += (sqrDataA1[i] + sqrDataA2[k]);
+                    dataRLocal[(i - startRow) * nVectors2 + k] *= coeff;
+                    if (dataRLocal[(i - startRow) * nVectors2 + k] < Math<algorithmFPType, cpu>::vExpThreshold())
+                    {
+                        dataRLocal[(i - startRow) * nVectors2 + k] = Math<algorithmFPType, cpu>::vExpThreshold();
+                    }
                 }
             }
-        }
-        daal::internal::Math<algorithmFPType, cpu>::vExp(nVectors1 * nVectors2, dataR, dataR);
+
+            daal::internal::Math<algorithmFPType, cpu>::vExp(numRowsInBlock * nVectors2, dataRLocal, dataRLocal);
+        });
     }
     else
     {
-        char uplo, trans;
+        char trans = 'T', notrans = 'N';
         algorithmFPType zero = 0.0, one = 1.0, two = 2.0;
-        uplo  = 'U';
-        trans = 'T';
-        if (isInParallel)
-        {
-            Blas<algorithmFPType, cpu>::xxsyrk(&uplo, &trans, (DAAL_INT *)&nVectors1, (DAAL_INT *)&nFeatures, &one, (algorithmFPType *)dataA1,
-                                               (DAAL_INT *)&nFeatures, &zero, dataR, (DAAL_INT *)&nVectors1);
-        }
-        else
-        {
-            Blas<algorithmFPType, cpu>::xsyrk(&uplo, &trans, (DAAL_INT *)&nVectors1, (DAAL_INT *)&nFeatures, &one, (algorithmFPType *)dataA1,
-                                              (DAAL_INT *)&nFeatures, &zero, dataR, (DAAL_INT *)&nVectors1);
-        }
-        for (size_t i = 0; i < nVectors1; i++)
-        {
-            for (size_t k = 0; k < i; k++)
+
+        services::Status retStat =
```

Comment:
Here you replaced xsyrk with xgemm for the case a1 == a2.
xgemm does 2X more computations than xsyrk because it computes full resulting matrix. In case a1 == a2, the result is a symmetric matrix. That means only half (i.e. upper or lower triangle) of the matrix can be computed, as it is done with xsyrk.

Reply:
I modified blocked gemm. So now only lower triangle of the matrix is computed